### PR TITLE
Refactor UnixSocket and UnixSocketChannel

### PR DESCRIPTION
**Pull request by marcuslinke**
@headius This is the second part of my PRs that covers UnixSocket related stuff. Please take a look. Thanks!

PS: This PR depends on https://github.com/jnr/jnr-enxio/pull/20
UPDATE: I've moved abstract base class from https://github.com/jnr/jnr-enxio/pull/20 into this PR for easier review. The `AbstractNativeSocketChannel` class should be moved to enxio later.

### CODE REVIEW

**Comment by felfert** [comment]
Hmm. This does not even compile locally?!
I get multiple errors regarding missing overrides of abstract methods. Just to compile , I had to make the following changes (of course the with missing method implementations):

@marcuslinke Did you actually compile this or am I missing something here?

```
diff --git a/src/main/java/jnr/unixsocket/UnixSocketChannel.java b/src/main/java/jnr/unixsocket/UnixSocketChannel.java
index c3f70c9..0a65979 100644
--- a/src/main/java/jnr/unixsocket/UnixSocketChannel.java
+++ b/src/main/java/jnr/unixsocket/UnixSocketChannel.java
@@ -20,20 +20,25 @@ package jnr.unixsocket;
 
 import java.io.IOException;
 import java.net.Socket;
+import java.net.SocketOption;
 import java.net.SocketAddress;
 import java.nio.ByteBuffer;
 import java.nio.channels.ClosedChannelException;
 import java.nio.channels.UnsupportedAddressTypeException;
+import java.nio.channels.SocketChannel;
 
 import jnr.constants.platform.Errno;
 import jnr.constants.platform.ProtocolFamily;
 import jnr.constants.platform.Sock;
 import jnr.constants.platform.SocketLevel;
-import jnr.constants.platform.SocketOption;
+import static jnr.constants.platform.SocketOption.SO_KEEPALIVE;
+import static jnr.constants.platform.SocketOption.SO_RCVTIMEO;
 import jnr.enxio.channels.AbstractNativeSocketChannel;
 import jnr.ffi.LastError;
 import jnr.ffi.byref.IntByReference;
 
+import java.util.Set;
+
 /**
  * A {@link java.nio.channels.Channel} implementation that uses a native unix
  * socket
@@ -161,6 +166,31 @@ public class UnixSocketChannel extends AbstractNativeSocketChannel {
 		}
 	}
 
+    @Override
+	public final SocketAddress getRemoteAddress() {
+        return getRemoteSocketAddress();
+    }
+    @Override
+	public final SocketAddress getLocalAddress() {
+        return localAddress;
+    }
+    @Override
+    public final <T> SocketChannel setOption(SocketOption<T> name, T value) {
+		throw new UnsupportedOperationException("not yet supported");
+    }
+    @Override
+    public SocketChannel bind(SocketAddress local) throws IOException {
+		throw new UnsupportedOperationException("not yet supported");
+    }
+    @Override
+    public Set<SocketOption<?>> supportedOptions() {
+		throw new UnsupportedOperationException("not yet supported");
+    }
+    @Override
+    public <T> T getOption(SocketOption<T> name) {
+		throw new UnsupportedOperationException("not yet supported");
+    }
+
 	public final UnixSocketAddress getRemoteSocketAddress() {
 		if (state != State.CONNECTED) {
 			return null;
@@ -233,23 +263,23 @@ public class UnixSocketChannel extends AbstractNativeSocketChannel {
 
 	public boolean getKeepAlive() {
 		int ret = Native.getsockopt(getFD(), SocketLevel.SOL_SOCKET,
-				SocketOption.SO_KEEPALIVE.intValue());
+				SO_KEEPALIVE.intValue());
 		return (ret == 1) ? true : false;
 	}
 
 	public void setKeepAlive(boolean on) {
 		Native.setsockopt(getFD(), SocketLevel.SOL_SOCKET,
-				SocketOption.SO_KEEPALIVE, on);
+				SO_KEEPALIVE, on);
 	}
 
 	public int getSoTimeout() {
 		return Native.getsockopt(getFD(), SocketLevel.SOL_SOCKET,
-				SocketOption.SO_RCVTIMEO.intValue());
+				SO_RCVTIMEO.intValue());
 	}
 
 	public void setSoTimeout(int timeout) {
 		Native.setsockopt(getFD(), SocketLevel.SOL_SOCKET,
-				SocketOption.SO_RCVTIMEO, timeout);
+				SO_RCVTIMEO, timeout);
 	}
 
 	@Override
```

**Comment by marcuslinke** [comment]
@felfert It does compile locally at least in my ide (eclipse) without errors. Need to check if it does also with maven at the command line this evening.

**Comment by felfert** [comment]
@marcuslinke Yep, I should probably have mention that: Using maven cmdline.
Also: **Watch out**: hedious has updated the version in jnr-enxio/pom.xml master branch to 0.14-SNAPSHOT so you probaly want to use something like 0.14.1-SNAPSHOT locally in order to ensure using the correct version. See my forked branch here: https://github.com/felfert/jnr-enxio/tree/PR20

**Comment by marcuslinke** [comment]
@felfert It compiles without any error at maven command line here. Maybe you use the wrong java version?

```
mvn -v
Apache Maven 3.2.5 (12a6b3acb947671f09b81f49094c53f426d8cea1; 2014-12-14T18:29:23+01:00)
Maven home: /Users/marcus/bin/apache-maven-3.2.5
Java version: 1.6.0_65, vendor: Apple Inc.
Java home: /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home
Default locale: de_DE, platform encoding: MacRoman
OS name: "mac os x", version: "10.9.5", arch: "x86_64", family: "mac"
```

Nevertheless there are some test errors I do not understand. Hope you can help with it...

**Comment by felfert** [comment]
I use the latest OpenJDK-8

**Comment by marcuslinke** [comment]
According to @headius the project must be compatible with java 6 until now.

**Comment by felfert** [comment]
Well, in pom.xml, the following should take care of that:
```
<plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>2.0.2</version>
        <configuration>
          <source>1.6</source>
          <target>1.6</target>
        </configuration>
</plugin>
```

**Comment by felfert** [comment]
This proves it:
[felfert@fritz jnr-unixsocket]$ file target/classes/jnr/unixsocket/Native.class 
target/classes/jnr/unixsocket/Native.class: compiled Java class data, version 50.0 (Java 1.6) [architecture=1792] [architecture=655371]

**Comment by marcuslinke** [comment]
In fact it doesn't. If you use OpenJDK8 as runtime for maven the pom declares source and target compatibility with java 6 but it compiles the sources with OpenJDK8 compiler and the appropriate libraries. Probably there are some extensions made to the nio classes in java 8 so you got compile errors then.

**Comment by felfert** [comment]
Well, even the .travis.yml specifies to build with openjdk6 and oraclejdk7. I specifically tested with oraclejdk7 here before I stated my comment. This resulted in exactly the same errors.

**Comment by marcuslinke** [comment]
OK. I can confirm the problems with oraclejdk 7 here too. So it compiles with java <= 6. Don't know how to deal with this problem. I think this means java 6 support must be dropped, right? We can't provide a solution that is compatible with java 6 and java 7.

**Comment by felfert** [comment]
Uh, thats unfortunate. I wonder, whether this is still true and why:

> According to @headius the project must be compatible with java 6 until now.

**Comment by felfert** [comment]
Otherwise I'd happily would drop java 6 support :)

**Comment by marcuslinke** [comment]
@felfert I agree but I fear we have to wait for @headius before we can proceed.

**Comment by marcuslinke** [comment]
Now it compiles and tests successfully on travis with java 6 at least. Ping @headius: How we should deal with java 6 / 7 incompatibilities?

File: src/main/java/jnr/enxio/channels/AbstractNativeSocketChannel.java [hunk 1]
```diff
@@ -27,102 +27,134 @@
...
+		int index = 0;
+
+		for (index = offset; index < length; index++) {
+			result += write(srcs[index]);
+		}
+
+		return result;
+	}
+
+	public SocketChannel shutdownInput() throws IOException {
```
Comment (felfert) [inline src/main/java/jnr/enxio/channels/AbstractNativeSocketChannel.java]: 1. Could you pleas move shutdownInput() and shutdownOutput() out of the base class.
  Reason:
  Those are SOCK_STREAM specific.
  I am currently working on Datagram channels in my personal fork and would like do inherit that from AbstractNativeSocketChannel.

2. For completeness, bind(), send() and receive() should probably be moved here too. (From my UnixDatagramChannel) see my personal branch at https://github.com/felfert/jnr-unixsocket/tree/datagram 

3. I really think we should have some sort of design meeting (virtually). Have you seen my PM?

**Comment by headius** [comment]
The Java 7 question is a tricky one. The only fix I know of is to have separate (abstract super, perhaps) classes for the different JDK versions and select the right one to construct at runtime.

**Comment by marcuslinke** [comment]
@headius Is it OK to drop Java 6 support then?

**Comment by headius** [comment]
It's probably ok. JRuby going forward does not support Java 6, and Java 6 has been EOL for many years now. @enebo might have concerns about 1.7 + UNIXSocket support, but I think it has fallen far behind at this point anyway.

**Comment by felfert** [comment]
I'd second that: Please drop Jave 6 support. This would make things **a lot** easier.

**Comment by felfert** [comment]
Hehe :) JRuby is going to drop even Java 7 - isn't it?

**Comment by headius** [comment]
@felfert At some point, yes :-) It remains to be seen if that will be in a 9.x release or not until JRuby 10k.

**Comment by felfert** [comment]
Are you on freenode #jnr right now?

File: src/main/java/jnr/enxio/channels/AbstractNativeSocketChannel.java (hunk 1 above)
Comment (marcuslinke) [inline src/main/java/jnr/enxio/channels/AbstractNativeSocketChannel.java]: @felfert I'm not sure if I understand you correctly. `AbstractNativeSocketChannel` is intended to be a `SocketChannel` ("A selectable channel for stream-oriented connecting sockets.") so its stream specific by nature I think.

...just invited you to docker-java gitter channel (https://gitter.im/docker-java) for further disscussions.

**Comment by marcuslinke** [comment]
Closing if favor of https://github.com/jnr/jnr-unixsocket/pull/35

**Pull request closed without merging**